modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/facet/CompositeFacetProcessor.java

```diff
@@ -54,28 +53,15 @@ public Optional<AggregationBuilder> processFacet(Facet facet) {
 		return facetProcessor.processFacet(facet);
 	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY,
-		target = "(&(class.name=*)(!(class.name=DEFAULT)))"
-	)
-	protected void setFacetProcessor(
-		FacetProcessor<SearchRequestBuilder> facetProcessor,
-		Map<String, Object> properties) {
-
-		String className = MapUtil.getString(properties, "class.name");
-
-		_facetProcessors.put(className, facetProcessor);
-	}
-
-	protected void unsetFacetProcessor(
-		FacetProcessor<SearchRequestBuilder> facetProcessor,
-		Map<String, Object> properties) {
-
-		String className = MapUtil.getString(properties, "class.name");
-
-		_facetProcessors.remove(className);
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
+			bundleContext,
+			(Class<FacetProcessor<SearchRequestBuilder>>)
+				(Class<?>)FacetProcessor.class,
+			"(&(class.name=*)(!(class.name=DEFAULT)))",
```

Comment:
@tinatian This class explicitly use OSGi filter to select services with `class.name` property and it's not `DEFAULT`, so I think using the `class.name` property is OK here.

By the way, providing the identifier via method makes it impossible for someone to reference one of the services of an interface. So for

```
public interface A
```

there could be a registry existing:

```
@Component
public class ARegistry
```

but there could also be some components referencing specific instance of A:

```
@Reference(target="(some.property=some.value)")
private A _a;
```

If we meet such case, do you want us to try to eliminate either the registry or the explicit references, or keep the property?

Reply:
@W2jx Do you have any example for the case above?